File: interviewbit-linked-lists-remove-nth-node-from-list-end.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param A : head node of linked list
    # @param B : integer
    # @return the head node in the linked list
    def removeNthFromEnd(self, A, B):
        length = 0
        current = A
        while current:
            length += 1
            current = current.next

        node_to_remove = max(length - B, 0)

        if node_to_remove == 0:
            return A.next

        current = A
        for i in range(length - B - 1):
            current = current.next

        current.next = current.next.next

        return A

File: interviewbit-linked-lists-remove-nth-node-from-list-end.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     public int val;
 *     public ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode a, int b) {
        int counter = 0;
        ListNode lead = a;
        ListNode follow = null;
        while (lead != null) {
            if (counter == b) {
                follow = a;
            }
            if (lead.next == null) {
                if (follow != null && follow.next != null) {
                    follow.next = follow.next.next;
                    return a;
                } else {
                    return a.next;
                }
            }
            lead = lead.next;
            counter += 1;
            if (follow != null) {
                follow = follow.next;
            }
        }
        return a.next;
    }
}
